### ISSUE TITLE
[Feature] Need traits to have proper generic collections

### ISSUE THREAD

**Issue by lsh** [modular/modular#113]
Minimal reproduction:
```mojo
from Vector import DynamicVector
fn foo(buf: DynamicVector[DynamicVector[UI32]]):
    for i in range(len(buf)):
        let b = buf[i]
        for j in range(len(b)):
            print(j)
```
Outputs:
```
error: Expression [9]:8:20: invalid call to '__getitem__': result cannot bind generic !mlirtype to memory-only type 'DynamicVector[SIMD[ui32, 1]]'
        let b = buf[i]
                ~~~^~~

/.modular/Kernels/mojo/Stdlib/Vector.mojo:417:5: function declared here
    fn __getitem__(self, i: Int) -> type:
    ^

error: Expression [9]:10:19: use of unknown declaration 'j'
            print(j)
```

**Comment by lattner** [comment]
Indeed, this is blocked on traits.

**Comment by Mogball** [comment]
This is a known issue. Closing this in favour of the roadmap because otherwise we will have tons of issues related to "no traits"

**Comment by lattner** [comment]
Maybe we should define a bug for each major missing feature and dupe against those?

**Comment by DayDun** [comment]
Sorry, but why exactly is this blocked on traits?

Smaller reproduction:
```py
struct A:
    pass

struct B[type: AnyType]:
    var value: type
    fn get_value(self) -> type:
        return self.value

fn f(x: B[A]):
    x.get_value()
```
```
error: Expression [13]:14:16: invalid call to 'get_value': result cannot bind generic !mlirtype to memory-only type 'A'
    x.get_value()
    ~~~~~~~~~~~^~
```

This does not error if `A` is `@register_passable`.

The `get_value` function itself typechecks, which I would assume would be where the issue occurs if it was a trait issue. Instead the error occurs during instantiation of the function.

Also I am curious exactly how traits are planned to be implemented, since all parameters are to my understanding just regular values at the moment, including types themselves.

Will regular values also be able to satisfy traits? 🤔

And yeah, creating issues for large (and small?) planned features sounds like a nice idea.

**Comment by Mogball** [comment]
That code is technically invalid too because the destructor for value is never called.

**Comment by lattner** [comment]
We need traits to reason about the associated methods of a type, specifically the dtor, copy ctor, move ctors, etc.
